service/raft/raft_group0_client.hh

```diff
@@ -107,7 +107,7 @@ public:
     // Call after `system_keyspace` is initialized.
     future<> init();
 
-    future<> add_entry(group0_command group0_cmd, group0_guard guard, seastar::abort_source* as);
+    future<> add_entry(group0_command group0_cmd, group0_guard guard, seastar::abort_source* as, std::optional<raft_timeout> timeout = std::nullopt);
```

Comment:
> If we want a default, I think raft_timeout{} is a better one.

Default value usually disables behaviour. It's more natural to expect that explicitly passed parameter adds behaviour, not removes it.

> I prefer debugging an explicit unexpected failure, than debugging a hang. If we have a failure we usually instantly know where to look. If we have a hang... it's much harder.

I strongly agree with this. Why not then take this a step further and declare 1 minute (or some other value) a group0 default operation timeout? We can still provide raft_timeout parameter with different semantics - if it's not null, its value field overrides the group level timeout parameter, if it's null - no timeout is applied.

Reply:
> Default value usually disables behaviour. It's more natural to expect that explicitly passed parameter adds behaviour, not removes it.

Philosophy. What's behavior here? Is waiting forever a behavior? Is waiting a limited amount of time a behavior? Which behavior is more of a behavior?

> I strongly agree with this. Why not then take this a step further and declare 1 minute (or some other value) a group0 default operation timeout? We can still provide raft_timeout parameter with different semantics - if it's not null, its value field overrides the group level timeout parameter, if it's null - no timeout is applied.

Didn't you just describe exactly what the PR already does?

The problem is if you call `add_entry` and forget to make a decision, i.e. don't pass `raft_timeout`, `nullopt` will be used, meaning that we will not have a timeout at all.

So the question is which outcome is better if the caller of `add_entry` forgets to make a decision.
- if the correct decision would be to wait forever, but instead we will timeout after 1 minute
- or if the correct decision would be to timeout, but instead we wait forever (hang)?

I think out of the two undesirable outcomes, the first one is better -- because it's much easier to debug.

Reply:
> Didn't you just describe exactly what the PR already does?

Nope, in the current PR version not passed raft_timeout parameter means don't apply timeout. I propose to change this to 'I don't care, use group-wide default'. If someone cares about timeouts, they can provide raft_timeout with explicit value, null for no timeout, time_point for specific timeout. This is currently not needed though if we take my approach (see below).

> I think out of the two undesirable outcomes, the first one is better -- because it's much easier to debug.

I agree.

We are basically talking about the same, the only question is what to do with background fibers. In any case, we would need to review all group0 usages in all of them. In your approach we would need to add `raft_timeout{}` to group0 calls, in my approach - add the `try/catch` to the top loop. I like my approach more because it doesn't require passing `raft_timeout{}` parameter down the call chains of the methods, which are used both in fibers and in user context (like in [auth commit](https://github.com/scylladb/scylladb/pull/17590/commits/da9d2b6ea0baedd5031756925f462fbbad5b9a5e), and there are several other such cases).

Reply:
> in my approach - add the try/catch to the top loop

So you're proposing to **always** have a timeout, and then deal with it with a try/catch + retry

My proposal is to have a timeout **by default**, but give the option to not have a timeout when needed (in background fibers), and pass it explicitly there

Reply:
>  In your approach we would need to add raft_timeout{} to group0 calls


Ah, you meant those group 0 calls that are being used in both contexts where we want and don't want a timeout

But I think there are fewer cases where we don't want timeout

Reply:
> In my approach we would not need to add raft_timeout{}. Because it would be the default
> I.e. change this:

I mean add std::nullopt to the call chains originating from background fibers.

Reply:
> have only top level timeout for entire operations (bootstrap a node, removenode, decomission etc).

We cannot do that. Those operations have unbounded runtime, based on the size of cluster / size of data etc.

It makes sense to have timeouts only for operations that should have a bound like schema changes or auth changes

Reply:
> I mean add std::nullopt to the call chains originating from background fibers.

Yes yes

but I actually think there are not  that many of them

For example in topology_coordinator there is... one :)

What are the others?

Reply:
> We cannot do that. Those operations have unbounded runtime, based on the size of cluster / size of data etc.
> 
> It makes sense to have timeouts only for operations that should have a bound like schema changes or auth changes

True.

Reply:
> For example in topology_coordinator there is... one :)

Ok sorry, I see two:
```cpp
    future<group0_guard> start_operation() {
        auto guard = co_await _group0.client().start_operation(&_as);
```

and
```cpp
    future<> update_topology_state(
            group0_guard guard, std::vector<canonical_mutation>&& updates, const sstring& reason) {
        try {
            rtlogger.info("updating topology state: {}", reason);
            rtlogger.trace("update_topology_state mutations: {}", updates);
            topology_change change{std::move(updates)};
            group0_command g0_cmd = _group0.client().prepare_command(std::move(change), guard, reason);
            co_await _group0.client().add_entry(std::move(g0_cmd), std::move(guard), &_as);
```

Reply:
I also think it's fine to have a timeout in the upgrade procedure fiber.

It is being triggered by a POST and waited on by a GET, so while it is not strictly "synchronous", the documented procedure is for the admin to monitor its progress by GET.

Reply:
> Nothing stops new ones from appearing.

And the same is true in the other direction i.e. those that should have a timeout but somebody will forget to pass it.

Again we're back to the dilemma presented above, what is worse -- new call will appear inside topo coordinator code that will timeout, OR new call will appear outside topo coordinator that will not timeout when it should

Reply:
> Again we're back to the dilemma presented above, what is worse -- new call will appear inside topo coordinator code that will timeout, OR new call will appear outside topo coordinator that will not timeout when it should

No, we're not. With both our approaches user will get timeout by default, but in my case they don't need additional plumbing for that. The plumbing, which adds more places where the wrong parameter could be passed.

Reply:
Then I don't understand. In your approach, if user wants to specify "no timeout", how do they do this? Or do you want to take away this possibility? (So it's not "timeout by default", it is "timeout always") And how do they specify a different timeout than the default?

Reply:
* My main point is that I want to reduce the amount of plumbing in the code;
  * I don’t want to introduce long chains where we need to pass raft_timeout{} from one function to another;
  * Simple solution - get rid of the reason why we need raft_timeout{} in one place but not in another
  * This means all group0 methods should have a default timeout; background fibers catch raft_operation_timeout_exception and retry at the top level
  * In the PR I proposed to introduce default timeout for raft_server_for_group, but it’s an implementation detail and not particularly important. We can introduce some global group0_timeout = 1 minute and use it in those places that communicate directly with raft_group_registry for group0 (mainly raft_group0_client + a few other)
  * The app level code (storage_service, migration_manager) won’t have to care about timeouts, except in try/catch for background fibers. We even don’t need raft_timeout parameter in raft_group0_client methods, they just always pass group0_timeout to the underlying raft_group0_registry api-s.

Reply:
As discussed on yesterday's daily -- let's proceed with the current approach.

Reply:
ok